### PR TITLE
enhancement(file source): Option to remove file after some time of reaching `eof`

### LINKED ISSUE #2187: Support file deletion after file has been read to the end
Our data source comes from a directory, an external program keeps writing new files to this directory. Currently, we use flume(Spooling Directory Source) to meet the needs.

Now I want to replace some flume with Vector, because flume is more heavyweight and more resource-intensive. 

But Vector does not support file deletion, I don’t want to use an external program to delete because I don't know if Vector has read to the end of the file.

### PR DESCRIPTION
**Pull request by ktff**
Closes #2187.

Adds a `remove_after` option which is a timeout from reaching `eof` after which file will be removed from filesystem, unless new data is written in the meantime. If not specified, files will not be removed.

### CODE REVIEW

File: lib/file-source/src/file_server.rs [hunk 1]
```diff
@@ -216,6 +217,23 @@ where
...
+                            if watcher.last_read_success().elapsed() >= grace_period {
+                                // Try to remove
+                                match remove_file(&watcher.path) {
+                                    Ok(()) => {
+                                        info!(message = "Log file deleted.", path = ?watcher.path);
+                                        watcher.set_dead();
+                                    }
+                                    Err(error) => {
+                                        // We will try again after some time.
+                                        warn!(message = "Failed deleting log file.",path = ?watcher.path, ?error);
```
Comment (lukesteensen) [inline lib/file-source/src/file_server.rs:231]: It's probably worth rate-limiting these.

**Review by lukesteensen — APPROVED** [review]

**Comment by ktff** [comment]
@binarylogic the website docs are ready for a review.

File: .meta/sources/file.toml.erb [hunk 2]
```diff
@@ -240,6 +240,17 @@ Instead of balancing read capacity fairly across all watched files, prioritize \
...
+[sources.file.options.remove_after]
+type = "uint"
+unit = "seconds"
+common = false
+examples = [0, 5, 60]
+description = """\
+Timeout from reaching `eof` after which file will be removed from filesystem, \
+unless new data is written in the meantime. \
+If not specified, files will not be removed.\
+"""
```
Comment (binarylogic) [inline .meta/sources/file.toml.erb:252]:
```suggestion
"""
warnings = [{visibility_level = "option", text = "Vector's process must have permission to delete files."}]

```

**Review by binarylogic — APPROVED** [review]
Nice work! Docs look great minus that one change.

**Pull request merged**